### PR TITLE
Refactor dask_cudf groupby to use apply_concat_apply

### PR DESCRIPTION
**Pull request by rjzamora**
Dask-cudf currently maintains a specialized groupby-aggregation code, this code is faster for cudf-based data than the upstream (`dask.dataframe`) code path. However, the custom implementation does not take advantage of Dask's `apply_concat_apply` function, even though the tree-reduction aspect of the algorithm is the same.

This PR refactors the dask_cudf groupby-aggregation code to use `apply_concat_apply`. This reduces the amount of code we will need to maintain in cudf, and should improve graph optimizations (like fusion).

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by rjzamora** [comment]
@charlesbluca - This PR will change the HLG layer(s) produced for a groupby aggregations. Do you expect this to break anything here or in dask-sql?

**Comment by charlesbluca** [comment]
I wouldn't expect it to change anything here, I might need to make some changes to https://github.com/rapidsai/cudf/pull/10853 but don't expect those to be too extensive.

I can pull this PR in with dask-sql to see if anything ends up breaking there

**Comment by charlesbluca** [comment]
Was able to check this PR against the dask-sql groupby tests, looks like the only test that's failing is [`test_groupby_split_every`](https://github.com/dask-contrib/dask-sql/blob/cdb8f9162015341f2d41e612ab369ca59344d730/tests/integration/test_groupby.py#L393-L432) which does a check on the exact number of keys in the HLG; we've since refactored this test to be less prone to upstream breakage in the datafusion branch, and it should be easy to incorporate those changes into `main`, so things should be good here 🙂

**Comment by rjzamora** [comment]
>Was able to check this PR against the dask-sql groupby tests

Thank you for taking the time to test this with dask-sql @charlesbluca  - That is super helpful!

**Comment by rjzamora** [comment]
**Update**: I used the proposed `local_cudf_groubpy.py` benchmark in https://github.com/rapidsai/dask-cuda/pull/979 to compare this PR to `branch-22.10`. The results confirm that this PR should not have a significant effect on performance (though a slight improvement is possible):

**This PR**

```
Groupby benchmark
--------------------------------------------------------------------------------
Use shuffle               | False
Output partitions         | 1
Input partitions          | 100
Sort Groups               | False
Rows-per-chunk            | 1000000
Unique-group ratio        | 0.001
Protocol                  | tcp
Device(s)                 | 0,1,2,3
Tree-reduction width      | 8
RMM Pool                  | True
Worker thread(s)          | 1
Data processed            | 2.24 GiB
Output size               | 3.05 MiB
Number of workers         | 4
================================================================================
Wall clock                | Throughput
--------------------------------------------------------------------------------
594.12 ms                 | 3.76 GiB/s
469.67 ms                 | 4.76 GiB/s
450.69 ms                 | 4.96 GiB/s
================================================================================
Throughput                | 4.43 GiB/s +/- 329.85 MiB/s
Bandwidth                 | 145.60 MiB/s +/- 20.04 MiB/s
Wall clock                | 504.82 ms +/- 63.61 ms
================================================================================
```
```
Groupby benchmark
--------------------------------------------------------------------------------
Use shuffle               | False
Output partitions         | 1
Input partitions          | 100
Sort Groups               | False
Rows-per-chunk            | 1000000
Unique-group ratio        | 0.01
Protocol                  | tcp
Device(s)                 | 0,1,2,3
Tree-reduction width      | 8
RMM Pool                  | True
Worker thread(s)          | 1
Data processed            | 2.24 GiB
Output size               | 30.52 MiB
Number of workers         | 4
================================================================================
Wall clock                | Throughput
--------------------------------------------------------------------------------
1.61 s                    | 1.39 GiB/s
1.29 s                    | 1.74 GiB/s
1.35 s                    | 1.66 GiB/s
================================================================================
Throughput                | 1.58 GiB/s +/- 92.59 MiB/s
Bandwidth                 | 181.52 MiB/s +/- 16.37 MiB/s
Wall clock                | 1.41 s +/- 140.13 ms
================================================================================
```
```
Groupby benchmark
--------------------------------------------------------------------------------
Use shuffle               | False
Output partitions         | 1
Input partitions          | 100
Sort Groups               | False
Rows-per-chunk            | 1000000
Unique-group ratio        | 0.1
Protocol                  | tcp
Device(s)                 | 0,1,2,3
Tree-reduction width      | 8
RMM Pool                  | True
Worker thread(s)          | 1
Data processed            | 2.24 GiB
Output size               | 305.16 MiB
Number of workers         | 4
================================================================================
Wall clock                | Throughput
--------------------------------------------------------------------------------
6.18 s                    | 370.31 MiB/s
6.56 s                    | 348.73 MiB/s
6.59 s                    | 347.12 MiB/s
================================================================================
Throughput                | 355.08 MiB/s +/- 5.98 MiB/s
Bandwidth                 | 168.96 MiB/s +/- 13.96 MiB/s
Wall clock                | 6.45 s +/- 187.88 ms
================================================================================
```

**`branch-22.10`**

```
Groupby benchmark
--------------------------------------------------------------------------------
Use shuffle               | False
Output partitions         | 1
Input partitions          | 100
Sort Groups               | False
Rows-per-chunk            | 1000000
Unique-group ratio        | 0.001
Protocol                  | tcp
Device(s)                 | 0,1,2,3
Tree-reduction width      | 8
RMM Pool                  | True
Worker thread(s)          | 1
Data processed            | 2.24 GiB
Output size               | 3.05 MiB
Number of workers         | 4
================================================================================
Wall clock                | Throughput
--------------------------------------------------------------------------------
668.32 ms                 | 3.34 GiB/s
556.43 ms                 | 4.02 GiB/s
510.05 ms                 | 4.38 GiB/s
================================================================================
Throughput                | 3.87 GiB/s +/- 262.52 MiB/s
Bandwidth                 | 175.48 MiB/s +/- 15.17 MiB/s
Wall clock                | 578.27 ms +/- 66.43 ms
================================================================================
```
```
Groupby benchmark
--------------------------------------------------------------------------------
Use shuffle               | False
Output partitions         | 1
Input partitions          | 100
Sort Groups               | False
Rows-per-chunk            | 1000000
Unique-group ratio        | 0.01
Protocol                  | tcp
Device(s)                 | 0,1,2,3
Tree-reduction width      | 8
RMM Pool                  | True
Worker thread(s)          | 1
Data processed            | 2.24 GiB
Output size               | 30.52 MiB
Number of workers         | 4
================================================================================
Wall clock                | Throughput
--------------------------------------------------------------------------------
1.56 s                    | 1.43 GiB/s
1.36 s                    | 1.64 GiB/s
1.41 s                    | 1.59 GiB/s
================================================================================
Throughput                | 1.55 GiB/s +/- 55.12 MiB/s
Bandwidth                 | 189.05 MiB/s +/- 12.99 MiB/s
Wall clock                | 1.44 s +/- 86.88 ms
================================================================================
```
```
Groupby benchmark
--------------------------------------------------------------------------------
Use shuffle               | False
Output partitions         | 1
Input partitions          | 100
Sort Groups               | False
Rows-per-chunk            | 1000000
Unique-group ratio        | 0.1
Protocol                  | tcp
Device(s)                 | 0,1,2,3
Tree-reduction width      | 8
RMM Pool                  | True
Worker thread(s)          | 1
Data processed            | 2.24 GiB
Output size               | 305.16 MiB
Number of workers         | 4
================================================================================
Wall clock                | Throughput
--------------------------------------------------------------------------------
6.19 s                    | 369.61 MiB/s
7.06 s                    | 323.97 MiB/s
6.67 s                    | 343.37 MiB/s
================================================================================
Throughput                | 344.65 MiB/s +/- 10.68 MiB/s
Bandwidth                 | 152.58 MiB/s +/- 19.05 MiB/s
Wall clock                | 6.64 s +/- 356.58 ms
================================================================================
```

**Review by galipremsagar — APPROVED** [review]

**Review by wence- — COMMENTED** [review]
Some minor nits

File: python/dask_cudf/dask_cudf/groupby.py [hunk 1]
```diff
@@ -514,10 +511,20 @@ def groupby_agg(
 
     # Deal with default split_out and split_every params
     if split_every is False:
-        split_every = ddf.npartitions
+        split_every = max(ddf.npartitions, 2)
```
[the remark this answers is not in the document]
Comment (rjzamora) [reply python/dask_cudf/dask_cudf/groupby.py]: Fair point - I broke out the `False` logic into a separate check (with a separate comment)

File: python/dask_cudf/dask_cudf/groupby.py [hunk 2]
```diff
@@ -618,26 +568,61 @@ def groupby_agg(
...
+        "sep": sep,
+        "str_cols_out": str_cols_out,
+        "aggs_renames": aggs_renames,
+    }
+
+    if shuffle:
+        # Temporary error until shuffle-based groupby is implemented
+        raise NotImplementedError(
+            "The shuffle option is not yet implemented in dask_cudf."
         )
```
[the remark this answers is not in the document]
Comment (rjzamora) [reply python/dask_cudf/dask_cudf/groupby.py]: Oops - forgot to do that earlier!

File: python/dask_cudf/dask_cudf/groupby.py [hunk 3]
```diff
@@ -484,6 +480,7 @@ def groupby_agg(
     sep="___",
```
[the remark this answers is not in the document]
Comment (rjzamora) [reply python/dask_cudf/dask_cudf/groupby.py:480]: This is indeed an "internal" implementation detail. However, it is technically possible for the original column names to conflict with the temporary-renaming strategy. Therefore, it is possible for the user to set a custom `sep` at the `GroupBy` level.

Perhaps it was overkill to make `sep` configurable, and we should remove it.  I'd be a bit suprised to learn that a real user has every **used** it. However, if we **do** decide to remove it, I would prefer to do that plumbing in a separate PR.

File: python/dask_cudf/dask_cudf/groupby.py [hunk 4]
```diff
@@ -742,31 +725,11 @@ def _groupby_partition_agg(
...
-            split_out_setup_kwargs={"cols": gb_cols},
-        )
-        del gb
-    return output
+    return gb
 
 
 @_dask_cudf_nvtx_annotate
-def _tree_node_agg(dfs, gb_cols, split_out, dropna, sort, sep):
+def _tree_node_agg(df, gb_cols, dropna, sort, sep):
```
[the remark this answers is not in the document]
Comment (rjzamora) [reply python/dask_cudf/dask_cudf/groupby.py:734]: Good call! Updated

File: python/dask_cudf/dask_cudf/groupby.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (wence-) [reply python/dask_cudf/dask_cudf/groupby.py:480]: I think a more foolproof way would be to internally generate unique names that are guaranteed not to collide (the dask version of this uses `tokenize` which is basically fine until someone starts putting md5 hashes in their column names)

[the remark this answers is not in the document]
Comment (rjzamora) [reply python/dask_cudf/dask_cudf/groupby.py:480]: >I think a more foolproof way would be to internally generate unique names that are guaranteed not to collide (the dask version of this uses tokenize which is basically fine until someone starts putting md5 hashes in their column names)

I agree that `tokenize` is much-more robust way to handle temporary column names

**Comment by rjzamora** [comment]
@gpucibot merge

**Pull request merged**